### PR TITLE
virtual gigacube + fsck setup

### PR DESCRIPTION
**Pull request by Golo300**
added virtual gigacube for simulating gigacube with pygame. Not all features of the real hardware are supported. Added multiplayer and fasck animation.

### CODE REVIEW

**Comment by Golo300** [comment]
works

File: faradaycage/gigacubeLibary/gigacube/__init__.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+from .gigacube import Gigacube
```
Comment (rubenhoenle) [inline faradaycage/gigacubeLibary/gigacube/__init__.py:1]: what is "faradaycage"? Could we have some words in the documentation about that? :smile:

File: faradaycage/gigacubeLibary/gigacube/gigacube.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+        K_RIGHT: False
+    }
+    buttons = [False, False]
+
+    def __init__(self):
+
+        self.led_states = [[[(random.randint(0, 255), random.randint(0, 255), random.randint(0, 255)) for _ in range(self.LED_MATRIX_SIZE)] for _ in range(self.LED_MATRIX_SIZE)] for _ in range(4)]
+        self.led_states_buff = [[[(random.randint(0, 255), random.randint(0, 255), random.randint(0, 255)) for _ in range(self.LED_MATRIX_SIZE)] for _ in range(self.LED_MATRIX_SIZE)] for _ in range(4)]
+        thread = threading.Thread(target=self.main, args=())
+        #thread.daemon = True                            # Daemonize thread
```
Comment (rubenhoenle) [inline faradaycage/gigacubeLibary/gigacube/gigacube.py:48]: remove commented code

File: faradaycage/gigacubeLibary/gigacube/gigacube.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+                    if event.key in self.pressed_keys:
+                        self.pressed_keys[event.key] = True
+
+                elif event.type == pygame.KEYUP:
+                    if event.key in [pygame.K_a, pygame.K_d]:
+                        self.rotate_angle_x = 0
+                    if event.key in [pygame.K_w, pygame.K_s]:
+                        self.rotate_angle_y = 0
+
+                    # virtuell input
```
Comment (rubenhoenle) [inline faradaycage/gigacubeLibary/gigacube/gigacube.py:175]:
```suggestion
                    # virtual input
```

File: faradaycage/gigacubeLibary/gigacube/gigacube.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+            fps = self.clock.get_fps()
+            fps_text = f"FPS: {int(fps)}"
+            self.drawText(10, 560, fps_text)
+            
+            self.drawcube()
+
+            glRotatef(self.rotate_angle_x, 0, 1, 0)
+            glRotatef(self.rotate_angle_y, 1, 0, 0)
+
+            #self.setColor(0, [(random.randint(0, 256), random.randint(0, 256), random.randint(0, 256)) if i < colorSize1 else None for i in range(self.LED_MATRIX_SIZE * self.LED_MATRIX_SIZE * 2)])
```
Comment (rubenhoenle) [inline faradaycage/gigacubeLibary/gigacube/gigacube.py:196]: remove commented code

File: faradaycage/gigacubeLibary/gigacube/hypervisor.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    def __init__(self):
+        raise RuntimeError('Call instance() instead')
+
+    @classmethod
+    def instance(cls):
+        if cls._instance is None:
+            
+            cls._instance = cls.__new__(cls)
+            cls._instance.cube = cls._instance.start_cube()
+            # Put any initialization here.
```
Comment (rubenhoenle) [inline faradaycage/gigacubeLibary/gigacube/hypervisor.py:15]: what's that comment for? Copy-pasta accident?

File: faradaycage/gigacubeLibary/gigacube/hypervisor.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+    def start_cube(self):
+
+        if self.cube != None:
+            return self.cube
+
+        import threading
+        from .gigacube import Gigacube
+
+        cube = Gigacube()
```
Comment (rubenhoenle) [inline faradaycage/gigacubeLibary/gigacube/hypervisor.py:26]: self.cube or cube?

File: faradaycage/neopixelLibary/neopixel/__pycache__/hypervisor.cpython-311.pyc
Comment (rubenhoenle) [inline faradaycage/neopixelLibary/neopixel/__pycache__/hypervisor.cpython-311.pyc:1]: gitignore `__pycache__`

File: flake.nix [hunk 7]
```diff
@@ -38,6 +100,42 @@
...
+
+                      # for developing the Visulizier
+                      pkgs.python3Packages.pygame
+                      pkgs.python3Packages.schedule
+                      pkgs.python3Packages.numpy
+                      pkgs.python39Packages.pyopengl
+                      ];
+
+        env = {
+          #AMPY_PORT = "/dev/ttyACM0";
```
Comment (rubenhoenle) [inline flake.nix:121]: why is that commented out?

File: lib/displaycontroller.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
+from machine import Pin, Timer
+from neopixel import NeoPixel
+from time import sleep_ms
+#import webserver
```
Comment (rubenhoenle) [inline lib/displaycontroller.py:4]: remove commented code

File: lib/nunchuck.py [hunk 9]
```diff
@@ -21,9 +21,16 @@
...
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 # SOFTWARE.
+import os
+virtuell = False
 
-import machine
-import time
+print("virtuell: ", virtuell)
+
+if virtuell:
```
Comment (rubenhoenle) [inline lib/nunchuck.py:29]:
```suggestion
if virtual:

```

File: lib/nunchuck.py [hunk 10]
```diff
@@ -37,18 +44,27 @@ class Nunchuck(object):
...
         mine as well, 400kHz does not)."""
+
+        
         self.i2c = i2c
         self.address = 0x52
         self.buffer = bytearray(b'\x00\x00\x00\x00\x00\x00')
         # There are two initialization sequences documented.
         #self.i2c.writeto(self.address, b'\x40\x00')
         #self.i2c.writeto(self.address, b'\x00')
+        if virtuell:
```
Comment (rubenhoenle) [inline lib/nunchuck.py:63]:
```suggestion
        if virtual:

```

File: lib/nunchuck.py [hunk 11]
```diff
@@ -62,6 +78,10 @@ def __init__(self, i2c, poll=True):
     def update(self):
         """Requests a sensor readout from the controller and receives the
         six data bits afterwards."""
+        if virtuell:
```
Comment (rubenhoenle) [inline lib/nunchuck.py:81]:
```suggestion
        if virtual:

```

File: lib/nunchuck.py [hunk 12]
```diff
@@ -70,12 +90,19 @@ def update(self):
 
     def __poll(self):
         """Poll the sensor readouts if necessary."""
+        if virtuell:
```
Comment (rubenhoenle) [inline lib/nunchuck.py:93]:
```suggestion
        if virtual:

```

File: lib/nunchuck.py [hunk 13]
```diff
@@ -70,12 +90,19 @@ def update(self):
...
+            self.update()
+            return
+        
         if time.ticks_diff(time.ticks_ms(), self.last_poll) > self.polling_threshold:
             self.update()
 
     def accelerator(self):
         """Retrieve the three axis of the last accelerometer reading.
         Returns a tuple of three elements: The x, y and z axis"""
+        if virtuell:
```
Comment (rubenhoenle) [inline lib/nunchuck.py:103]:
```suggestion
        if virtual:

```

File: lib/nunchuck.py [hunk 14]
```diff
@@ -85,47 +112,72 @@ def buttons(self):
         """Returns a tuple representing the states of the button C and Z
         (in this order). The values are True if the corresponding button
         is being pressed and False otherwise."""
+        if virtuell:
```
Comment (rubenhoenle) [inline lib/nunchuck.py:115]:
```suggestion
        if virtual:

```

File: lib/nunchuck.py [hunk 15]
```diff
@@ -85,47 +112,72 @@ def buttons(self):
...
+            self.__poll()
+            return self.h.getButtons()
         self.__poll()
         return (not (self.buffer[5] & 2 == 2), not (self.buffer[5] & 1 == 1))
 
     def joystick(self):
         """Get the X and Y positions of the thumb joystick. For both axis
         a value of 0 means left, ~136 center and 255 right position."""
+        if virtuell:
```
Comment (rubenhoenle) [inline lib/nunchuck.py:124]:
```suggestion
        if virtual:

```

File: lib/nunchuck.py [hunk 16]
```diff
@@ -85,47 +112,72 @@ def buttons(self):
...
+            raise NotImplementedError
         self.__poll()
         return (self.buffer[0], self.buffer[1])
 
     def joystick_left(self):
         """Returns True if the joystick is being held to the left side."""
+        if virtuell:
```
Comment (rubenhoenle) [inline lib/nunchuck.py:131]:
```suggestion
        if virtual:

```

File: lib/nunchuck.py [hunk 17]
```diff
@@ -85,47 +112,72 @@ def buttons(self):
...
+            self.__poll()
+            return self.h.isLeft()
+        
         self.__poll()
         return self.buffer[0] < 55
 
     def joystick_right(self):
         """Returns True if the joystick is being held to the right side."""
+        if virtuell:
```
Comment (rubenhoenle) [inline lib/nunchuck.py:140]:
```suggestion
        if virtual:

```

File: lib/nunchuck.py [hunk 18]
```diff
@@ -85,47 +112,72 @@ def buttons(self):
...
+            self.__poll()
+            return self.h.isRight()
         self.__poll()
         return self.buffer[0] > 200
 
     def joystick_up(self):
         """Returns True if the joystick is being held to the upper side."""
+        if virtuell:
```
Comment (rubenhoenle) [inline lib/nunchuck.py:148]:
```suggestion
        if virtual:

```

File: lib/nunchuck.py [hunk 19]
```diff
@@ -85,47 +112,72 @@ def buttons(self):
...
+            self.__poll()
+            return self.h.isUp()
         self.__poll()
         return self.buffer[1] > 200
 
     def joystick_down(self):
         """Returns True if the joystick is being held to the lower side."""
+        if virtuell:
```
Comment (rubenhoenle) [inline lib/nunchuck.py:156]:
```suggestion
        if virtual:

```

File: lib/nunchuck.py [hunk 20]
```diff
@@ -85,47 +112,72 @@ def buttons(self):
...
+            self.__poll()
+            return self.h.isDown()
         self.__poll()
         return self.buffer[1] < 55
 
     def joystick_center(self):
         """Returns True if the joystick is not moved away fromthe center
         position."""
+        if virtuell:
```
Comment (rubenhoenle) [inline lib/nunchuck.py:165]:
```suggestion
        if virtual:

```

File: lib/nunchuck.py [hunk 21]
```diff
@@ -85,47 +112,72 @@ def buttons(self):
...
+            self.__poll()
+            return self.h.isCenter()
         self.__poll()
         return self.buffer[0] > 100 and self.buffer[0] < 155 and self.buffer[1] > 100 and self.buffer[1] < 155
 
     def joystick_x(self):
         # Returns "normalized" x axis values
+        if virtuell:
```
Comment (rubenhoenle) [inline lib/nunchuck.py:173]:
```suggestion
        if virtual:

```

File: lib/nunchuck.py [hunk 22]
```diff
@@ -85,47 +112,72 @@ def buttons(self):
...
+            raise NotImplementedError
         self.__poll()
         return (self.buffer[0] >> 2) - 34
 
     def joystick_y(self):
         # Returns "normalized" x axis values
+        if virtuell:
```
Comment (rubenhoenle) [inline lib/nunchuck.py:180]:
```suggestion
        if virtual:

```

File: main.py [hunk 23]
```diff
@@ -148,55 +123,105 @@ def checkCookies(self):
...
+        last_one = None # if only one left this is the one
+
+        for player in self.players:
+            if player.alive:
+                still_alive += 1
+                last_one = player
+
+        if still_alive == 1:
+            self.win_animation(timer, last_one)
+        if still_alive == 0:
```
Comment (rubenhoenle) [inline main.py:182]: use else if here

File: main.py [hunk 24]
```diff
@@ -242,9 +277,16 @@ def nunchuck_update(nunchuck: Nunchuck, player_id: int, gamelogic: GameLogic):
     #        a[0], a[1], a[2],
```
Comment (rubenhoenle) [inline main.py:277]: remove commented code

File: main.py [hunk 25]
```diff
@@ -228,9 +263,9 @@ def nunchuck_update(nunchuck: Nunchuck, player_id: int, gamelogic: GameLogic):
 
     
     b = nunchuck.buttons()
-    if((b[0] or b[1]) and len(gamelogic.players) == 0): # condition seems to be true for some seconds after the start of the game
+    if((b[0] or b[1]) and gamelogic.idle): # condition seems to be true for some seconds after the start of the game
+        print("here")
         gamelogic.startGame()
-    #print( b[0], b[1] )
 
 #_thread.start_new_thread(blink, gamelogic)
```
Comment (rubenhoenle) [inline main.py:270]: remove commented code

File: main.py [hunk 26]
```diff
@@ -242,9 +277,16 @@ def nunchuck_update(nunchuck: Nunchuck, player_id: int, gamelogic: GameLogic):
...
     #        b[0], b[1]
     #        ))
-
+0
     #sleep_ms(2)
+
+idlescreen = IdleScreen(gamelogic.left, gamelogic.display_controller)
+
 while True:
     #webserver.webserver_hook(gamelogic)
```
Comment (rubenhoenle) [inline main.py:286]: remove commented code

**Review by rubenhoenle — CHANGES_REQUESTED** [review]
see above

File: faradaycage/gigacubeLibary/gigacube/hypervisor.py (hunk 5 above)
Comment (Golo300) [reply faradaycage/gigacubeLibary/gigacube/hypervisor.py:15]: Yes copy pasta

File: faradaycage/gigacubeLibary/gigacube/hypervisor.py [hunk 27]
```diff
@@ -0,0 +1,46 @@
...
+            cls._instance.cube = cls._instance.start_cube()
+            # Put any initialization here.
+        return cls._instance
+
+    def start_cube(self):
+
+        if self.cube != None:
+            return self.cube
+
+        import threading
```
[the remark this answers is not in the document]
Comment (Golo300) [reply faradaycage/gigacubeLibary/gigacube/hypervisor.py:23]: I remember it being on purpose but I don't remember why... let me think

File: faradaycage/gigacubeLibary/gigacube/__init__.py (hunk 1 above)
Comment (Golo300) [reply faradaycage/gigacubeLibary/gigacube/__init__.py:1]: wrote README

File: faradaycage/gigacubeLibary/gigacube/hypervisor.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (Golo300) [reply faradaycage/gigacubeLibary/gigacube/hypervisor.py:22]: it is important

**Review by rubenhoenle — APPROVED** [review]

File: faradaycage/gigacubeLibary/gigacube/hypervisor.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (rubenhoenle) [reply faradaycage/gigacubeLibary/gigacube/hypervisor.py:22]: Sounds like pfusch but well it works...

**Pull request merged**